osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/AttributeAddressHelperService.java

```diff
@@ -0,0 +1,248 @@
+        default:
+            throw new ProtocolAdapterException(String.format("periodtype %s not supported", periodType));
+        }
+    }
+
+    private List<AttributeAddress> createScalerUnitForInterval() {
+        final List<AttributeAddress> scalerUnit = new ArrayList<>();
+        scalerUnit.add(new AttributeAddress(CLASS_ID_REGISTER, OBIS_CODE_INTERVAL_IMPORT_SCALER_UNIT,
+                ATTRIBUTE_ID_SCALER_UNIT));
+        scalerUnit.add(new AttributeAddress(CLASS_ID_REGISTER, OBIS_CODE_INTERVAL_EXPORT_SCALER_UNIT,
+                ATTRIBUTE_ID_SCALER_UNIT));
+        return scalerUnit;
+    }
+
+    private List<AttributeAddress> createScalerUnitForMonth() {
+        final List<AttributeAddress> scalerUnit = new ArrayList<>();
+        scalerUnit.add(new AttributeAddress(CLASS_ID_REGISTER, OBIS_CODE_MONTHLY_DAILY_IMPORT_RATE_1_SCALER_UNIT,
+                ATTRIBUTE_ID_SCALER_UNIT));
+        scalerUnit.add(new AttributeAddress(CLASS_ID_REGISTER, OBIS_CODE_MONTHLY_DAILY_IMPORT_RATE_2_SCALER_UNIT,
+                ATTRIBUTE_ID_SCALER_UNIT));
+        scalerUnit.add(new AttributeAddress(CLASS_ID_REGISTER, OBIS_CODE_MONTHLY_DAILY_EXPORT_RATE_1_SCALER_UNIT,
+                ATTRIBUTE_ID_SCALER_UNIT));
+        scalerUnit.add(new AttributeAddress(CLASS_ID_REGISTER, OBIS_CODE_MONTHLY_DAILY_EXPORT_RATE_2_SCALER_UNIT,
+                ATTRIBUTE_ID_SCALER_UNIT));
+        return scalerUnit;
+    }
+
+    private SelectiveAccessDescription getSelectiveAccessDescription(final PeriodTypeDto periodType,
+            final DateTime beginDateTime, final DateTime endDateTime, final boolean isSelectingValuesSupported) {
+        /*
+         * List of object definitions to determine which of the capture objects
+         * to retrieve from the buffer.
+         */
+        final List<DataObject> objectDefinitions = new ArrayList<>();
+        if (isSelectingValuesSupported) {
+            this.addSelectedValues(periodType, objectDefinitions);
+        }
+        final DataObject selectedValues = DataObject.newArrayData(objectDefinitions);
+
+        final DataObject accessParameter = this.dlmsHelperService.getAccessSelectionTimeRangeParameter(beginDateTime,
+                endDateTime, selectedValues);
+
+        return new SelectiveAccessDescription(ACCESS_SELECTOR_RANGE_DESCRIPTOR, accessParameter);
+    }
+
+    private void addSelectedValues(final PeriodTypeDto periodType, final List<DataObject> objectDefinitions) {
+
+        switch (periodType) {
+        case INTERVAL:
+            // empty objectDefinitions is ok, since all values are applicable,
+            // hence selective access is not applicable
+            break;
+        case DAILY:
+            this.addSelectedValuesForDaily(objectDefinitions);
+            break;
+        case MONTHLY:
+            this.addSelectedValuesForMonthly(objectDefinitions);
+            break;
+        default:
+            throw new AssertionError("Unknown PeriodType: " + periodType);
```

Comment:
I see three switch statements on periodType in this class. This shouts "USE POLYMORPHISM!". I suggest creating an interface (e.g. 'PeriodHandler') with the methods getProfileBuffer (if that's the name you chose for the extracted method I suggested above),  getScalerUnit, addSelectedValues in it. Then create three separate implementations of the PeriodHandler interface: IntervalPeriodHandler, DayPeriodHandler, MonthPeriodHandler and put the relevant logic in each class. You could even use the PeriodType enum as an undercover factory to give an instance of PeriodHandler matching the enum value.

Notice there is a fourth switch statement hiding away in GetPeriodicMeterReadsCommandExecutor.processNextPeriodicMeterReads. Is there any possibility to rake that one in with the refactoring described above?

Reply:
I suggest to leave it for now as it is. When we refactor the OBIS codes configuration, then it would not be needed anymore.